Refactor InternalAbstractGraphDatabase

rickardoberg opened the pull request:
Split IAGDB into a factory and a facade. The factory instantiates 3 modules (platform, edition, datasource), and subclasses then specifies how each edition is instantiated.

rickardoberg committed:
Refactor InternalAbstractGraphDatabase
Split IAGDB into a factory and a facade. The factory instantiates 3 modules (platform, edition, datasource), and subclasses then specifies how each edition is instantiated.

Disable online backup for store copy

thobe commented:
I have a feeling that the `refactorings.patch` file should not have been added.

chrisvest commented:
The `data/dbms/auth` file also looks like it was added by mistake.

rickardoberg commented:
@thobe @chrisvest fixed both, thanks for catching that!

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -0,0 +1,584 @@
+    @Override
+    public IndexManager index()
+    {
+        // TODO: txManager.assertInUnterminatedTransaction();
```
Why is this not done?

rickardoberg replied:
Good question, I don't know. All the method implementations are just copied verbatim from IAGDB. That line was from a commit made by Jakub in July last year. Maybe we should add it? Not sure why it was commented out, and if we can maybe safely add it back.

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -0,0 +1,584 @@
+    private static final long MAX_NODE_ID = IdType.NODE.getMaxValue();
+    private static final long MAX_RELATIONSHIP_ID = IdType.RELATIONSHIP.getMaxValue();
+
+    private boolean initialized = false;
```
I can't quite see this flag serving a purpose. It is only assigned in the constructor, so it could be final. But the only cases when that assignment wouldn't take place are cases when the object wouldn't be published (because of exceptions in the constructor) so it does in fact serve no purpose.

Perhaps it should be traded for a flag that helps making shutdown idempotent?

rickardoberg replied:
This is another one that was simply copied. It is not set in the constructor though, but rather the init() method. In 3.0 I hope we can change this to be constructor though, which would be better than the init() method.

But replacing it with shutdown flag is definitely an option.

thobe replied:
No. I'm silly. that's not a constructor, that's a method called `init`... I'm still wondering if `shutdown()` shouldn't set this field to false.

thobe commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -0,0 +1,584 @@
+
+    public PlatformModule platformModule;
+    public EditionModule editionModule;
+    public DataSourceModule dataSourceModule;
```
These module fields doesn't appear to be used (except for platform module which is used for accessing the `config`, but that could be extracted to a separate field in the init method just like the other components of these objects).

I'd propose removing these fields from here. The module types still serve a purpose, as vectors of their respective fields, but after that purpose have been served, I don't see the point of keeping those objects around.

rickardoberg replied:
The main reason I left them is if we want to have tests access them instead of using DependencyResolver. In general it would be nice if we could cut down on DependencyResolver usage, as it is a bit overused as it is. With the module fields it will be much easier to do "Find usages" when we make refactorings of services and whatnot. Makes sense?

rickardoberg replied:
We could also opt for removing all the other fields, and have accesses go through module refs, as there should be no performance overhead for doing so.

thobe replied:
Makes sense. I didn't notice that they were public.

We could have all access go through the module refs, although I do think there would be a (tiny) performance overhead of doing so, since it is one additional indirection.

lutovich replied:
Imho public fields are very bad and replacing hackish `DependencyResolver` (which we at least have control over) with those is not a good idea.

rickardoberg replied:
The reason is because with DependencyResolver it is non-trivial to know what services are used by what code. With the public fields we can do "Find usages" during refactorings. Making them final is a good point though.

rickardoberg committed:
Clean up how we reference JUnit+Hamcrest+Mockito in POM files. Make sure we avoid -all deps that screw up classpath.

digitalstain committed:
Fixed masking of startup exception with shutdown exception in IAGD

This is a forward port of #4458 from 2.2, adjusted to the new
 realities of the refactoring around IAGD to Facade and friends.

cleishm commented:
Squash?

rickardoberg commented:
@cleishm no, because the two commits do very different things. The first is for the main refactoring, and the second updates all POM files with respect to how we use JUnit, Hamcrest, and Mockito (to avoid classpath hell issues).

cleishm commented:
I was referring to the "Merge pull request #2" commit. Seems odd that a PR
would contain a merge from a wholly different repo, and that'll look very
weird in our history.

On Wed, Apr 22, 2015 at 11:23 PM, Rickard Öberg notifications@github.com
wrote:

> @cleishm https://github.com/cleishm no, because the two commits do very
> different things. The first is for the main refactoring, and the second
> updates all POM files with respect to how we use JUnit, Hamcrest, and
> Mockito (to avoid classpath hell issues).
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/neo4j/neo4j/pull/4466#issuecomment-95457105.

rickardoberg commented:
@cleishm yup, understand what you mean, fixed

cleishm commented on advanced/server-advanced/src/main/java/org/neo4j/server/advanced/AdvancedNeoServer.java:
```diff
@@ -32,12 +32,12 @@
     }
 
-    public AdvancedNeoServer( ConfigurationBuilder configurator, InternalAbstractGraphDatabase.Dependencies dependencies, LogProvider logProvider )
+	public AdvancedNeoServer( ConfigurationBuilder configurator, CommunityFacadeFactory.Dependencies dependencies, LogProvider logProvider)
```
you might like to run the code-formatter over these.

cleishm commented on community/kernel/src/main/java/org/neo4j/kernel/extension/KernelExtensions.java:
```diff
@@ -63,7 +64,10 @@ public void init() throws Throwable
-                life.add( kernelExtensionFactory.newKernelExtension( configuration) );
+                Lifecycle dependency = kernelExtensionFactory.newKernelExtension( configuration );
+                if (dependency == null)
+                    throw new NullPointerException( kernelExtensionFactory.toString()+" returned a null KernelExtension" );
```
`Objects.requireNonNull( dependency, kernelExtensionFactory + " provided a null KernelExtension" )` ?

tinwelint commented:
You seem to have gotten another unrelated commit into this PR (related to what @cleishm mentioned before). This one: `Fixed masking of startup exception with shutdown exception in IAGD` should not be in here. It seems to come from #4458

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/EditionModule.java:
```diff
@@ -0,0 +1,74 @@
+ */
+public abstract class EditionModule
+{
+    public IdGeneratorFactory idGeneratorFactory;
```
Public non-final fields???

tinwelint replied:
I don't think that's as bad as it looks, since that's internal classes. Compare that to a setter and a getter which effectively is the same thing. The difference is that you can control access to it, although since this is internal stuff we can change this at any time and the compiler will help us complete such a change.

cleishm replied:
There shouldn't be a setter for that - if there is, then something is wrong in the way code is structured. It should only be set from within the class, preferably only during construction.

Having a getter is nice for controlling access and later refactoring.

rickardoberg replied:
I can add final, that's fine. Adding getters wouldn't add anything beyond that.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacadeFactory.java:
```diff
@@ -0,0 +1,208 @@
+ * which create all the specific services needed to run a graph database.
+ *
+ * It is abstract in order for subclasses to specify their own {@link org.neo4j.kernel.impl.factory.EditionModule} implementations. Subclasses also have to set the edition name
+ * in overriden version of {@link #newFacade(java.util.Map, org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.Dependencies, GraphDatabaseFacade)}, which is used for logging and similar.
```
Maybe it is better to define one more method `String getName()` instead of putting a requirement to modify `Map` parameter?

rickardoberg replied:
I had that at one point. Various points of code wants to know what edition is running though, so having that available beyond just the factory is more useful.

lutovich replied:
Well, than I think it should be possible to define such method and make default impl of `newFacade` that puts value from that method into map. Than we'll only need to require `super.newFacade` call.

tinwelint commented:
Personally I think the amount of introduced checkstyle violations is a bit on the high side. How can we improve this moving further? I'm not militant about fixing those before merge, but it'd make me happier.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/GraphDatabaseDependencies.java:
```diff
@@ -26,14 +26,21 @@
+    public static GraphDatabaseDependencies newDependencies(GraphDatabaseFacadeFactory.Dependencies deps)
+    {
+        return new GraphDatabaseDependencies( deps.monitors(), deps.userLogProvider(), Iterables.toList(deps.settingsClasses
+                () ), Iterables.toList(deps.kernelExtensions() ), Iterables.toList(deps.executionEngines() ) );
```
I dislike the point at this line is broken into two, surely automatic line breaking should be able to be more context-aware. Also different style of space before/after parentheses.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/CommunityFacadeFactory.java:
```diff
@@ -0,0 +1,49 @@
+    extends GraphDatabaseFacadeFactory
+{
+    @Override
+    public GraphDatabaseFacade newFacade( Map<String, String> params, Dependencies dependencies, GraphDatabaseFacade
```
I find it harder to read code that have line breaks in weird places, such as between type and variable name. Your IDE should be able to be smarter about this,

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/DataSourceModule.java:
```diff
@@ -0,0 +1,391 @@
+            @Override
+            public GraphDatabaseService getGraphDatabase()
+            {
+                // TODO This should be wrapped as well
```
Should it be wrapped or is it an old comment?

rickardoberg replied:
Most of this is copied verbatim from IAGDB. If we have issues with IAGDB code, then lets deal with that after this is merged.

cleishm commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacadeFactory.java:
```diff
@@ -0,0 +1,208 @@
+     * @return
+     */
+    public GraphDatabaseFacade newFacade( Map<String, String> params, final Dependencies dependencies,
+                                          final GraphDatabaseFacade graphDatabaseFacade )
```
Is this meant to be private? (Factory methods don't take already constructed objects of the type they produce)

rickardoberg replied:
The only reason this method is there at all is because of EmbeddedGraphDatabase and other IAGDB subclasses, which (as per devteam email) need to remain until 3.0 for compatibility reasons. We can remove this later.

cleishm commented on community/kernel/src/test/java/org/neo4j/kernel/GraphDatabaseFacadeFactoryTest.java:
```diff
@@ -0,0 +1,126 @@
+                PlatformModule module = new PlatformModule( mock( Map.class ), mock( Dependencies.class, RETURNS_MOCKS ),
+                        mock( GraphDatabaseFacade.class ) );
+                module.life = mock( LifeSupport.class );
+                module.storeDir = mock( File.class );
```
It would be better to pass in mocks during construction, rather than poking them into what are essentially internal properties.

cleishm commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacadeFactory.java:
```diff
@@ -0,0 +1,208 @@
+
+            platform.life.start();
+        }
+        catch ( final Throwable throwable )
```
Should this really catch `Throwable`? That seems excessive.

rickardoberg replied:
This is copied verbatim from IAGDB. If we have issues with that, let's change after the PR is done. This is mainly a refactoring, not a refactoring and touchup.

cleishm commented:
This doesn't seem to address `HighlyAvailableGraphDatabaseFactory`

rickardoberg commented:
@cleishm as per devteam email, HighlyAvailableGraphDatabase needs to remain, and since that now uses this mechanism, it is addressed. Once HAGD and other IAGDB subclasses are removed, we can change HAGDBF as well.
